Консольное приложение №1. Первая проектная работа.

PhilipMaslov1 opened the pull request.

PhilipMaslov1 committed:
Консольное приложение №1. Первая проектная работа.

kgornostaeva commented on src/main/java/Main.java:
```diff
@@ -1,6 +1,44 @@
+            Calculator.askGoodPrice(good, scanner); // Уточняем стоимость товара.
+            Calculator.addGoodToBasket(good); // Добавляем товар в корзину.
+
+            System.out.println("Добавить ещё один товар?"); // Уточняем, хочет ли пользователь что-то ещё добавить в корзину.
```
Круто было бы подсказать пользователю, как он может завершить сценарий, чтобы он не гадал

PhilipMaslov1 replied:
Добавил подсказку.

kgornostaeva commented on src/main/java/Main.java:
```diff
@@ -1,6 +1,44 @@
-        System.out.println("Hello world!");
+
+        int numberOfPerson; // Количество участников.
+        Scanner scanner = new Scanner(System.in);
```
Чтобы сканнер не расходовал ресурсы системы, важно закрывать его в конце использования, чтобы не допускать утечек памяти. Используй метод close()

PhilipMaslov1 replied:
Работу с консолью вынес в отдельный класс - ConsoleReader. Туда же добавил метод закрытия сканера, который вызывается по завершении main.

kgornostaeva commented on src/main/java/Main.java:
```diff
@@ -1,6 +1,44 @@
     public static void main(String[] args) {
-        System.out.println("Hello world!");
+
+        int numberOfPerson; // Количество участников.
```
Рекомендую вынести весь этот код в отдельный класс, который будет считывать данные из ввода, сохранять их и вызывать методы других классов для обработки данных (подсчета, форматирования).
В нем публичными нужно оставить только те методы, вызов которых нужно производить снаружи и для которых изменение порядка этих вызовов не сломает логику. Сейчас наиболее подходящим классом является Calculator (несмотря на название)

PhilipMaslov1 replied:
Вынес в отдельный класc ConsoleReader.

kgornostaeva commented on src/main/java/Good.java:
```diff
@@ -0,0 +1,6 @@
+public class Good {
+    String name;
+    double price;
+    String listOfGoods = ""; // Корзина товаров (список покупок)
```
Корзина товаров и общая сумма лучше хранить отдельно от конкретного товара.  Иначе каждый товар будет внутри себя содержать список товаров и занимать лишнюю память. Ты можешь вынести их в отдельные поля основного класса, отвечающего за функционал подсчета и взаимодействия с пользователем, либо создать отдельный класс "Basket" и уже его экземпляр сохранить в основном классе, но не в классе товара

PhilipMaslov1 replied:
Разнёс продукт и счёт в разные классы (product и bill).

kgornostaeva commented on src/main/java/Good.java:
```diff
@@ -0,0 +1,6 @@
+public class Good {
```
Рекомендую переименовать в Product, так понятнее. Насколько знаю, в английском языке слово good употребляется в значении "товар" только во множественном числе - goods. Возможно я не права, но Product считывается лучше и является общепринятым обозначением

PhilipMaslov1 replied:
Переименовал в Product.

kgornostaeva commented on src/main/java/Formatter.java:
```diff
@@ -0,0 +1,37 @@
+
+    // Метод возвращает слово "рубль" с окончанием, в зависимости от суммы.
+    public static String endOfRoubles (double sum){
+        switch ((int) Math.floor(sum) - 10 * ((int) Math.floor(sum) / 10)) {
```
Здесь два раза происходит приведение к int и округление одного и того же числа. Лучше сохранить его в переменную и использовать ее в операциях

PhilipMaslov1 replied:
Спасибо. Поправил.

kgornostaeva commented on src/main/java/Formatter.java:
```diff
@@ -0,0 +1,37 @@
+    // Метод возвращает слово "рубль" с окончанием, в зависимости от суммы.
+    public static String endOfRoubles (double sum){
+        switch ((int) Math.floor(sum) - 10 * ((int) Math.floor(sum) / 10)) {
+            case 1:
```
Если в твоей версии джавы это доступно, то ты можешь использовать enhanced switch, он более компактный:
```
return switch ((int) Math.floor(sum) - 10 * ((int) Math.floor(sum) / 10)) {
            case 1 -> "рубль.";
            case 2, 3, 4 -> "рубля.";
            case 5, 6, 7, 8, 9, 0 -> "рублей.";
            default -> "руб.";
        };
```
Доступен с Java 12

PhilipMaslov1 replied:
Привёл к виду encanced switch.

kgornostaeva commented on src/main/java/Formatter.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    // Метод выводит сообщение с долей каждого участника.
+    public static void resultOfDividing (Good good, int numberOfPerson) {
```
Кажется, странным, что метод для подсчета попал в класс для форматирования. Кажется, ему место в классе Calculator

PhilipMaslov1 replied:
Поправил. Перенёс в класс Calculator.

kgornostaeva commented on src/main/java/Formatter.java:
```diff
@@ -0,0 +1,37 @@
+        double sumPerPerson = Math.round(good.sum / numberOfPerson * 100) / 100.0;
+        double remOfDivision = Math.round((good.sum - sumPerPerson * numberOfPerson) * 100) / 100.0;
+
+        // Если сумма не может быть разделена поровну, считаем сколько должен заплатить каждый участник в максимально равных долях.
```
Можно разделить этот метод на два: один будет считать результаты, а второй выводить их на печать

PhilipMaslov1 replied:
Разделил. Все операции с числами вынес в класс Calculator и из других классов обращаюсь через методы.

kgornostaeva commented on src/main/java/Formatter.java:
```diff
@@ -0,0 +1,37 @@
+
+        // Если сумма не может быть разделена поровну, считаем сколько должен заплатить каждый участник в максимально равных долях.
+        if (remOfDivision == 0)
+            System.out.printf("Счёт можно разделить поровну.\nКаждый должен заплатить по %.2f %s", sumPerPerson, Formatter.endOfRoubles(sumPerPerson));
```
В конце финальной фразы рекомендую добавить перевод строки, чтобы вывод программы не прилип к следующему за ним системному промпту

PhilipMaslov1 replied:
Готово.

kgornostaeva commented on src/main/java/Calculator.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Scanner;
+
+public class Calculator {
```
Думаю, что данный класс можно расширить и включить в него методы и поля, которые сейчас находятся в Main, а конкретные математические операции вынести в отдельный класс Calculator. Этот тогда лучше будет переименовать, например, в SplitBill, BillSplitter или нечто подобное, отражающее его цель

PhilipMaslov1 replied:
Внёс изменения. Создал класс Bill, описывающий счёт и его основные параметры
;
Работу со считыванием данных вынес в класс ConsoleReader;
По твоему  совету мат. операции перенс в класс Calculator;

kgornostaeva commented on src/main/java/Calculator.java:
```diff
@@ -0,0 +1,33 @@
+
+    // Метод, уточняющий стоимость товара
+    public static void askGoodPrice (Good good, Scanner scanner) {
+        System.out.println("Введите стоимость в формате \"руб.коп\":");
```
Круто, что подсказываешь пользователю, в каком формате вводить данные!

PhilipMaslov1 replied:
Спасибо )

kgornostaeva commented on src/main/java/Calculator.java:
```diff
@@ -0,0 +1,33 @@
+            scanner.next();
+            System.out.println("Стоимость нужно ввести в формате \"руб.коп\":");
+        }
+        good.price = Math.round(scanner.nextDouble() * 100) / 100d;
```
Вот здесь может произойти переполнение, о котором я писала в комментарии к проекту. Почитать о переполнении в java можно тут https://www.baeldung.com/java-overflow-underflow

PhilipMaslov1 replied:
Прочитал про переполнение. Подскажи, пожалуйста, нужно ли добавить к операциям проверки на переполнение? Условиями успешности задачи это не предусмотрено, а чтобы корректно добавить их в код, мне нужно доплонительное время, чтобы исследовать работу с ними.

kgornostaeva replied:
Обсудили с твоим наставником, что все-таки для зачета данного проекта нет необходимости обрабатывать переполнение. Так что проверю проект без учета этого замечания)

kgornostaeva commented on src/main/java/Main.java:
```diff
@@ -1,6 +1,44 @@
+
+            System.out.println("Добавить ещё один товар?"); // Уточняем, хочет ли пользователь что-то ещё добавить в корзину.
+            if (scanner.next().equalsIgnoreCase("Завершить")) {
+                Calculator.printBasket(good); // Печатаем состав корзины и её стоимость.
```
Общий комментарий: поменьше комментариев ))  Кажется, что сейчас у тебя в проекте излишнее количество комментариев, часть из которых объясняет то, что и так понятно.
Это, конечно, холиварный вопрос, но лично я рекомендую писать код так, чтобы он был самообъясняемым, самодокументируемым Для этого важно правильно подбирать названия методов и полей, не всовывать внутрь методов неявную логику.
Если очень хочется добавить комментарии, то сделай это только для публичных методов и полей, которые могут использовать другие разработчики, если им понадобится твой класс.

PhilipMaslov1 committed:
Консольное приложение №1. Первая проектная работа.

Учел, за исключением одного, все комментарии из https://github.com/PhilipMaslov1/Java-Module-Project-YP/pull/1

ToDo: нужно добавить обработку переполнений при считывании числовых значений и выполнении математических операций.

PhilipMaslov1 committed:
Консольное приложение №1. Первая проектная работа.

Изменения:
- учтены замечания PR https://github.com/PhilipMaslov1/Java-Module-Project-YP/pull/1;
- исправлена ошибка с возможностью ввести сумму товара < 0.00;
- исправлнеа ошибка на ввод суммы товара в виде очень маленьких значений (0.0001);
- исправил ошибку с неверным окончанием слова "рубль." для некоторых сумм.

ToDo: нужно добавить обработку переполнений при считывании числовых значений и выполнении математических операций.
С ревьюером и настаником согласовано влитие без учёта этого требования.

PhilipMaslov1 commented:
@kgornostaeva внёс множество правок, поправил ошибки и постарался учесть твои комментарии. Посмотри, пожалуйста. Задание отправил на проверку.
